include/rocksdb/options.h

```diff
@@ -475,6 +475,33 @@ class ReplicationLogListener {
   // the database needs to re-apply all replication log records since
   // DB::GetPersistedReplicationSequence() (non-inclusive).
   virtual std::string OnReplicationLogRecord(ReplicationLogRecord record) = 0;
+
+  // Unlike the other log event, kMemtableSwitch needs to be split into two
+  // phases:
+
+  // Phase1): Get the replication sequence and assign it to all the
+  // switched memtablesx
+  // Phase2): Write the kMemtableSwitch event to log (with the
+  // replication sequence)
+  //
+  // TODO(wei): note that there is race condition with current implementation!
```

Comment:
@karenl1 @igorcanadi I would like to get your opinions on this bug. I think we might need to change the design on the log recovery to fix this issue.

Reply:
I'm not sure I understand why "`kMemtableSwitch` needs to be split into two phases"?

Reply:
I chatted with Igor offline and figured out. This is not necessary. 

Quick summary:

The `kMemtableSwitch` is a little bit different from the other events. We need to get a replication_sequence before memtable switch (this replication_sequence number will be assigned to the switched memtables and persist into manifest file when we write the corresponding `kManifestWrite`) and collect all the next_log_num after we switch all the memtables. The replication_sequence and the next_log_nums will then be serialized into the replication log as `kMemtableSwitch`.

Let's call the replication_sequence we generate for memtable switch as `mem_table_rs`, replication_sequence we generate for `kManifestWrite` as `manifest_write_rs`. Since memtable switch and write to manifest can happen in different threads, so it's possible to have following events in the log:
```
MemtableSwitch (mem_switch_rs = 5)
ManifestWrite (mem_switch_rs = 5, manifest_rs=7)
MemtableSwitch (mem_switch_rs = 6)
ManifestWrite (mem_switch_rs = 6, manifest_rs = 8)
```

I was concerned about this since ManifestWrite with manifest_rs = 7 is put in front of MemtableSwitch with mem_switch_rs=6 in the log. But turns out we really don't care about manifest_rs and the ordering. As long as the corresponding kManifestWrite happens after the kMemtableSwitch, we are good

Reply:
Gotcha, that makes sense, thanks for explaining!